### ISSUE TITLE
With the parameter $PreserveFQDN on the logged-in host is not the full host(FQDN)

### ISSUE THREAD

**Issue by Roarca** [apache/logging-log4j2#1740]
## Description

When on the client side you activate the **$PreserveFQDN** property on **/etc/rsyslog.conf** , the logs do not include the FQDN and only include the host 

## Configuration
My rsyslog.conf configuration (everything is default except the parameter)
```txt
$PreserveFQDN on
```
The /etc/hosts file
```txt
127.0.0.1 host.app.example host
```

And the log4j2 configuration
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Configuration>
	<Appenders>
		<Console name="LogToConsole" target="SYSTEM_OUT">
			<PatternLayout pattern="%d{HH:mm:ss.SSS} [%t] %-5level %logger{36} - %msg%n" />
		</Console>
		<Syslog name="RFC5424" format="RFC5424" host="localhost" port="514" protocol="TCP" appName="app" includeMDC="true" facility="LOCAL0" enterpriseNumber="18060" newLine="true" messageId="app-ID" mdcId="mdc" id="app-MDCID" connectTimeoutMillis="1000" immediateFail="true" reconnectionDelayMillis="5000">
			<LoggerFields>
				<KeyValuePair key="thread" value="%t" />
				<KeyValuePair key="priority" value="%p" />
				<KeyValuePair key="category" value="%c" />
				<KeyValuePair key="exception" value="%ex" />
				<KeyValuePair key="message" value="%m" />
			</LoggerFields>
		</Syslog>
	</Appenders>
	<category name="org.springframework.beans">
		<priority value="info" />
	</category>
	<Loggers>
		<Logger name="foo.example.ErrorService" additivity="false">
			<AppenderRef ref="RFC5424" />
			<AppenderRef ref="LogToConsole" />
		</Logger>
	</Loggers>
</Configuration>
```

**Version:** [Log4j version] 

> 2.17.0

**Operating system:** [OS and version]

> SUSE Linux Enterprise Server 12 SP3

**JDK:** [JDK distribution and version]  

> OpenJDK 64-Bit Server VM (build 25.292-b10, mixed mode)

## Logs

```log
2023-08-24 14:01:49.569 host app[4539] [status] Starting...
```

## Reproduction

Activating the **$PreserveFQDN** on parameter and having in **/etc/hosts** an FQDN longer than the normal host should be enough, but ask any questions you may have

**Comment by ppkarwasz** [comment]
@Roarca,
Thanks for the report. The bug is easily reproducible if the system call `gethostname()` returns an unqualified host name.

We use [`InetAddress#getLocalHost()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/InetAddress.html#getLocalHost()) to retrieve the local hostname, which calls `gethostname()`, performs a lookup for the **fully qualified** hostname and **ignores** the result (cf. [source code](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/unix/native/libnet/Inet4AddressImpl.c#L198)).

This is why you see just `host` as hostname.

As a workaround, you can change the hostname of the machine:
```sh
hostname host.app.example
```
It is not the UNIX way, but most services should work without any problems. You can also use [`uts_namespaces`](https://man7.org/linux/man-pages/man7/uts_namespaces.7.html) to isolate the hostname of your app from the rest of the system.

**Comment by drTr0jan** [comment]
Hi, @ppkarwasz. Where have you found `gethostname()` ignores the result?
`InetAddress.getHostName()` returns updated `holder().getHostName()` (if previously was null).
`InetAddress.getHostName()` differs from `InetAddress.getCanonicalHostName()` by mandatory reverse name lookup only.
`InetAddress.getHostName()` returns FQDN too.

I think, it will be great to return `useFQDN` config feature for using `InetAddress.getHostName()`.

**Comment by ppkarwasz** [comment]
Hi @drTr0jan,

Sorry, my earlier comment was ambiguous: by `gethostname()` (all lowercase) I meant the [POSIX C function](https://man7.org/linux/man-pages/man2/gethostname.2.html), not a Java method.

On Linux [`InetAddress.getLocalHost()`](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/net/InetAddress.html#getLocalHost()) does this:

1. `impl.getLocalHostName()` calls `gethostname()`: that is *all* it does, so you get the kernel's UTS `nodename`, traditionally unqualified.
2. That name goes to `lookupAllHostAddr()`, which calls [`getaddrinfo()`](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html) with the `AI_CANONNAME` hint, so the C library does resolve a canonical name.
3. The JNI code then stamps the **input** string onto every returned address, discarding `res->ai_canonname`:

```c
setInetAddress_hostName(env, iaObj, host);
```

([`Inet4AddressImpl.c#L198`](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.base/unix/native/libnet/Inet4AddressImpl.c#L198); `Inet6AddressImpl.c` does the same at lines 362 and 384, and that is the implementation actually used on a dual-stack host.)

You are right that `InetAddress.getHostName()` performs a reverse [`getnameinfo()`](https://man7.org/linux/man-pages/man3/getnameinfo.3.html) lookup, but only when the name is not already known:

```java
String getHostName(boolean check) {
    if (holder().getHostName() == null) {
        holder().hostName = InetAddress.getHostFromNameService(this, check);
    }
    return holder().getHostName();
}
```

Step 3 already filled that field in, so the lookup never happens. `getHostName()` does return an FQDN for addresses obtained other ways (`getByAddress()`, or from a socket), just not for this one.

**Comment by drTr0jan** [comment]
@ppkarwasz, Piotr, thanks for the fast answer.

> On Linux [InetAddress.getLocalHost()](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/net/InetAddress.html#getLocalHost()) does this:
> impl.getLocalHostName() calls gethostname(): that is all it does, so you get the kernel's UTS nodename, traditionally unqualified.

Hm... But on Darwin and FreeBSD `gethostname(3)` returns **qualified** hostname.

> On Linux [InetAddress.getLocalHost()](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/net/InetAddress.html#getLocalHost()) does this:
> ... 
>    That name goes to lookupAllHostAddr(), which calls [getaddrinfo()](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html) with the AI_CANONNAME hint, so the C library does resolve a canonical name.

Am I to understand, `InetAddress.getLocalHost()` after `impl.getLocalHostName()` does next..?

1. `InetAddress.getLocalHost()` calls `getAllByName0()` for reverse lookup the name (`local`).
2. `InetAddress.getAllByName0()` calls `NameServiceAddresses.get()` for reverse lookup the name (`host`).
3. `NameServiceAddresses.get()` calls `getAddressesFromNameService()' for get address for the name (`host`) from a NS.
4. `InetAddress.getAddressesFromNameService()` calls `lookupByName()` for lookup by the name (`host`).
5. `PlatformResolver.lookupByName()` calls `lookupAllHostAddr()`, which calls [getaddrinfo()](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html) with the AI_CANONNAME hint, so the C library does resolve a canonical name.

Is it correct?

**Comment by ppkarwasz** [comment]
> Hm... But on Darwin and FreeBSD `gethostname(3)` returns **qualified** hostname.

Sorry, you are right: traditions vary. Anyway the result of this call can be qualified or unqualified. For the OP the name was unqualified.

> Am I to understand, `InetAddress.getLocalHost()` after `impl.getLocalHostName()` does next..?
> 
> 1. `InetAddress.getLocalHost()` calls `getAllByName0()` for reverse lookup the name (`local`).
> 2. `InetAddress.getAllByName0()` calls `NameServiceAddresses.get()` for reverse lookup the name (`host`).
> 3. `NameServiceAddresses.get()` calls `getAddressesFromNameService()' for get address for the name (`host`) from a NS.
> 4. `InetAddress.getAddressesFromNameService()` calls `lookupByName()` for lookup by the name (`host`).
> 5. `PlatformResolver.lookupByName()` calls `lookupAllHostAddr()`, which calls [getaddrinfo()](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html) with the AI_CANONNAME hint, so the C library does resolve a canonical name.
> 
> Is it correct?

Yes, the result is similar to `InetAddress.getAllByName()` with the name given by the call above.  However, no canonicalization occurs: the JDK does request `AI_CANONNAME`, but it does not use it. It replaces whatever the C library returns with the host name obtained above.

**TL;DR** If `gethostname()` returns `foo`, then `InetAddress.getLocalHost().getHostName()` also returns `foo`, which was the root cause of this issue.

**Comment by drTr0jan** [comment]
Hi, @ppkarwasz.
Thx for detailed answer. I've investigated the flow of `InetAddress.getLocalHost()` in detail. And I believe that using  `lookupAllHostAddr()` as SHOULD canonicalizational is incorrect. I think that AI_CANONNAME was been added to the `lookupAllHostAddr()` by mistake. And I agree with @vy, who opened https://bugs.openjdk.org/browse/JDK-8392344

Besides, I think that using canonicalization for lookup hostname it's not always good idea (but Linux still uses this case). There's two cases where it works incorrect:
1. Server (no containers) has a hostname, but haven't the hostname in DNS or hosts(5). In that case we have `host` as `localhost`, which is incorrect.
2. Container (not Linux, ex. FreeBSD jail) has a hostname, but has the hostname  in hosts(5) as a second string (after the localhost). In that case we have `host` as `localhost` after 2.24.0. And have `host` as correct hostname before 2.24.0.
